src/client/client.go

```diff
@@ -423,42 +428,29 @@ func NewOnUserMachine(prefix string, options ...Option) (*APIClient, error) {
 
 	// create new pachctl client
 	var fw *PortForwarder
-	addr, cfgOptions, err := getUserMachineAddrAndOpts(context)
+	pachdAddress, cfgOptions, err := getUserMachineAddrAndOpts(context)
 	if err != nil {
 		return nil, err
 	}
-	if addr == "" {
-		addr = fmt.Sprintf("0.0.0.0:%s", DefaultPachdNodePort)
+	if pachdAddress == nil {
+		pachdAddress = &grpcutil.DefaultPachdAddress
 		fw = portForwarder()
 	}
 
-	client, err := NewFromAddress(addr, append(options, cfgOptions...)...)
+	client, err := NewFromAddress(pachdAddress.Hostname(), append(options, cfgOptions...)...)
 	if err != nil {
 		if strings.Contains(err.Error(), "context deadline exceeded") {
-			// port always starts after last colon, but net.SplitHostPort returns an
-			// error on a hostport without a colon, which this might be
-			port := ""
-			if colonIdx := strings.LastIndexByte(addr, ':'); colonIdx >= 0 {
-				port = addr[colonIdx+1:]
-			}
 			// Check for errors in approximate order of helpfulness
-			if port != "" && port != DefaultPachdPort && port != DefaultPachdNodePort {
+			if pachdAddress.IsUnusualPort() {
 				return nil, fmt.Errorf("could not connect (note: port is usually "+
-					"%d or %d, but is currently set to \"%d\"--is this right?): %v",
-					DefaultPachdNodePort, DefaultPachdPort, port, err)
-			}
-			isLoopback := strings.HasPrefix(addr, "0.0.0.0") || strings.HasPrefix(addr, "127.0.0.1") || strings.HasPrefix(addr, "[::1]") || strings.HasPrefix(addr, "localhost")
-			if fw == nil && isLoopback {
-				return nil, fmt.Errorf("could not connect (note: address %q looks "+
-					"like loopback, check that 'pachctl port-forward' is running): %v",
-					addr, err)
-			}
-			if port == "" {
-				return nil, fmt.Errorf("could not connect (note: address %q does not "+
-					"seem to be host:port): %v", addr, err)
+					"%d or %d, but is currently set to %d - is this right?): %v", grpcutil.DefaultPachdNodePort, grpcutil.DefaultPachdPort, pachdAddress.Port, err)
+			} else if fw == nil && pachdAddress.IsLoopback() {
+				return nil, fmt.Errorf("could not connect (note: address %s looks "+
+					"like loopback, try unsetting it): %v",
+					pachdAddress.Qualified(), err)
 			}
 		}
-		return nil, fmt.Errorf("could not connect to pachd at %q: %v", addr, err)
+		return nil, fmt.Errorf("could not connect to pachd at %s: %v", pachdAddress.Qualified(), err)
```

Comment:
I'm not too bothered either way, but why change the formatting directive? Do the quotes look bad?

In the past I've favored `%q` over `%s` because it shows accidental leading and trailing spaces in user input, but I'd be happy to start doing the opposite

Reply:
TBH, I think I just changed this from muscle memory. Happy to stick to `%q`.